File: test/lookup/lookup.cpp
#include <lookup/entry.hpp>
#include <lookup/input.hpp>
#include <lookup/lookup.hpp>

#include <stdx/bitset.hpp>
#include <stdx/utility.hpp>

#include <catch2/catch_test_macros.hpp>

#include <array>
#include <cstdint>

TEST_CASE("a lookup with no entries", "[lookup]") {
    constexpr auto lookup =
        lookup::make(CX_VALUE(lookup::input<std::uint32_t>{}));
    STATIC_REQUIRE(lookup[0u] == 0u);
    STATIC_REQUIRE(lookup[42u] == 0u);
}

TEST_CASE("a lookup with no entries and non-zero default", "[lookup]") {
    constexpr auto lookup =
        lookup::make(CX_VALUE(lookup::input<std::uint32_t>{5u}));
    STATIC_REQUIRE(lookup[0u] == 5u);
    STATIC_REQUIRE(lookup[42u] == 5u);
}

namespace {
template <auto N> using bitset = stdx::bitset<N, std::uint32_t>;
}

TEST_CASE("a linear lookup with non-integer values", "[lookup]") {
    constexpr auto lookup = lookup::linear_search_lookup<25>::make(
        CX_VALUE(lookup::input<int, bitset<32>, 2>{
            bitset<32>{},
            std::array{
                lookup::entry{42, bitset<32>{stdx::place_bits, 0}},
                lookup::entry{89, bitset<32>{stdx::place_bits, 12, 3}}}}));

    STATIC_REQUIRE(lookup[0] == bitset<32>{});
    STATIC_REQUIRE(lookup[42] == bitset<32>{stdx::place_bits, 0});
    STATIC_REQUIRE(lookup[89] == bitset<32>{stdx::place_bits, 12, 3});
}

TEST_CASE("a lookup with some entries", "[lookup]") {
    constexpr auto lookup = lookup::make(CX_VALUE(lookup::input<int, int, 3>{
        0, std::array{lookup::entry{0, 13}, lookup::entry{6, 42},
                      lookup::entry{89, 10}}}));
    STATIC_REQUIRE(lookup[0] == 13);
    STATIC_REQUIRE(lookup[1] == 0);
    STATIC_REQUIRE(lookup[5] == 0);
    STATIC_REQUIRE(lookup[6] == 42);
    STATIC_REQUIRE(lookup[7] == 0);
    STATIC_REQUIRE(lookup[88] == 0);
    STATIC_REQUIRE(lookup[89] == 10);
    STATIC_REQUIRE(lookup[90] == 0);
}

TEST_CASE("a lookup with some entries and non-zero default", "[lookup]") {
    constexpr auto lookup = lookup::make(CX_VALUE(lookup::input<int, int, 3>{
        5, std::array{lookup::entry{1, 13}, lookup::entry{6, 42},
                      lookup::entry{89, 10}}}));
    STATIC_REQUIRE(lookup[0] == 5);
    STATIC_REQUIRE(lookup[1] == 13);
    STATIC_REQUIRE(lookup[5] == 5);
    STATIC_REQUIRE(lookup[6] == 42);
    STATIC_REQUIRE(lookup[7] == 5);
    STATIC_REQUIRE(lookup[88] == 5);
    STATIC_REQUIRE(lookup[89] == 10);
    STATIC_REQUIRE(lookup[90] == 5);
}

TEST_CASE("a lookup with some sequential entries", "[lookup]") {
    constexpr auto lookup =
        lookup::make(CX_VALUE(lookup::input<std::uint32_t, std::uint32_t, 10>{
            1u, std::array{lookup::entry{0u, 13u}, lookup::entry{1u, 42u},
                           lookup::entry{2u, 10u}, lookup::entry{3u, 76u},
                           lookup::entry{4u, 25u}, lookup::entry{5u, 82u},
                           lookup::entry{6u, 18u}, lookup::entry{7u, 87u},
                           lookup::entry{8u, 55u}, lookup::entry{9u, 11u}}}));
    STATIC_REQUIRE(lookup[0] == 13);
    STATIC_REQUIRE(lookup[1] == 42);
    STATIC_REQUIRE(lookup[2] == 10);
    STATIC_REQUIRE(lookup[3] == 76);
    STATIC_REQUIRE(lookup[4] == 25);
    STATIC_REQUIRE(lookup[5] == 82);
    STATIC_REQUIRE(lookup[6] == 18);
    STATIC_REQUIRE(lookup[7] == 87);
    STATIC_REQUIRE(lookup[8] == 55);
    STATIC_REQUIRE(lookup[9] == 11);
    STATIC_REQUIRE(lookup[10] == 1);
    STATIC_REQUIRE(lookup[11] == 1);
    STATIC_REQUIRE(lookup[12] == 1);
    STATIC_REQUIRE(lookup[13] == 1);
    STATIC_REQUIRE(lookup[14] == 1);
    STATIC_REQUIRE(lookup[15] == 1);
    STATIC_REQUIRE(lookup[4'000'000'000u] == 1);
}

File: test/lookup/pbt/lookup.py
from hypothesis import strategies as st, given, settings, event, assume
import subprocess

def unpack(l):
    return ", ".join([str(i) for i in l])

uint16s = st.integers(min_value = 0, max_value = (1 << 16) - 1)


def lookup_make(lookup_type, key_type, value_type, default, entries):
    entries = [f"lookup::entry<{key_type}, {value_type}>{{{k}, {v}}}" for k, v in entries]
    return f"""
        {lookup_type}::make(CX_VALUE(lookup::input<{key_type}, {value_type}, {len(entries)}>{{
            {default},
            std::array<lookup::entry<{key_type}, {value_type}>, {len(entries)}>{{ {unpack(entries)} }}
        }}))
    """

@st.composite
def lookup_inputs(draw, keys=uint16s, values=uint16s, default=uint16s, min_size=0, max_size=10):
    entries = draw(st.lists(st.tuples(keys, values), min_size=min_size, max_size=max_size, unique_by=lambda x: x[0]))
    default = draw(default)
    return (default, entries)

pseudo_pext_lookups = st.sampled_from([
    "lookup::pseudo_pext_lookup<>",
    "lookup::pseudo_pext_lookup<true, 1>",
    "lookup::pseudo_pext_lookup<true, 2>",
    "lookup::pseudo_pext_lookup<true, 3>",
    "lookup::pseudo_pext_lookup<true, 4>"
])

@settings(deadline=50000)
@given(
    pseudo_pext_lookups, 
    lookup_inputs(min_size=2, max_size=12), 
    st.lists(uint16s, min_size=10, max_size=1000, unique=True)
)
def test_lookup(compile, t, l, extras):
    default, entries = l

    lookup_model = {k: v for k, v in entries}
    lookup = lookup_make(t, "std::uint16_t", "std::uint16_t", default, entries)

    check_keys = set(lookup_model.keys()).union(set(extras))

    static_asserts = "\n".join(
        [f"static_assert(lookup[{k}] == {lookup_model.get(k, default)});" for k in check_keys] 
    )

    runtime_checks = " &&\n".join(
        [f"(lookup[{k}] == {lookup_model.get(k, default)})" for k in check_keys] 
    )

    out = compile(f"""
        #include <lookup/entry.hpp>
        #include <lookup/input.hpp>
        #include <lookup/lookup.hpp>

        #include <stdx/utility.hpp>

        #include <array>
        #include <cstdint>

        int main() {{
            [[maybe_unused]] constexpr auto lookup = {lookup};
            {static_asserts}

            bool const pass = {runtime_checks};

            return pass ? 0 : 1;
        }}
    """)

    result = subprocess.run([out], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    assert result.returncode == 0